drivers/spi/spi_context.h

```diff
@@ -0,0 +1,239 @@
+				    struct k_poll_signal *async)
+{
+	k_sem_take(&ctx->lock, K_FOREVER);
+
+#ifdef CONFIG_POLL
+	if (async) {
+		ctx->async = async;
+	}
+#endif
+}
+
+static inline void spi_context_release(struct spi_context *ctx, int status)
+{
+	if (!status && (ctx->config->operation & SPI_LOCK_ON)) {
+		return;
+	}
+
+#ifdef CONFIG_POLL
+	if (!ctx->async || status) {
+		k_sem_give(&ctx->lock);
+	}
+#else
+	k_sem_give(&ctx->lock);
+#endif
+}
+
+static inline void spi_context_unlock_unconditionally(struct spi_context *ctx)
+{
+	if (!k_sem_count_get(&ctx->lock)) {
+		k_sem_give(&ctx->lock);
+	}
+}
+
+static inline void spi_context_wait_for_completion(struct spi_context *ctx)
+{
+#ifdef CONFIG_POLL
+	if (!ctx->async) {
+		k_sem_take(&ctx->sync, K_FOREVER);
+	}
+#else
+	k_sem_take(&ctx->sync, K_FOREVER);
+#endif
+}
+
+static inline void spi_context_complete(struct spi_context *ctx, int status)
+{
+#ifdef CONFIG_POLL
+	if (!ctx->async) {
+		k_sem_give(&ctx->sync);
+	} else {
+		k_poll_signal(ctx->async, status);
+
+		if (!(ctx->config->operation & SPI_LOCK_ON)) {
+			k_sem_give(&ctx->lock);
+		}
+	}
+#else
+	k_sem_give(&ctx->sync);
+#endif
+}
+
+static inline void spi_context_cs_configure(struct spi_context *ctx)
+{
+	if (ctx->config->cs) {
+		gpio_pin_configure(ctx->config->cs->gpio_dev,
+				   ctx->config->cs->gpio_pin, GPIO_DIR_OUT);
+		gpio_pin_write(ctx->config->cs->gpio_dev,
+			       ctx->config->cs->gpio_pin, 1);
+	}
+}
+
+static inline void spi_context_cs_control(struct spi_context *ctx, bool on)
+{
+	if (ctx->config->cs) {
+		if (on) {
+			gpio_pin_write(ctx->config->cs->gpio_dev,
+				       ctx->config->cs->gpio_pin, 0);
+			k_busy_wait(ctx->config->cs->delay);
+		} else {
+			if (ctx->config->operation & SPI_HOLD_ON_CS) {
+				return;
+			}
+
+			k_busy_wait(ctx->config->cs->delay);
+			gpio_pin_write(ctx->config->cs->gpio_dev,
+				       ctx->config->cs->gpio_pin, 1);
+		}
+	}
+}
+
+static inline void spi_context_buffers_setup(struct spi_context *ctx,
+					     const struct spi_buf **tx_bufs,
+					     struct spi_buf **rx_bufs,
+					     uint8_t dfs)
+{
+	SYS_LOG_DBG("tx_bufs %p (%p) - rx_bufs %p (%p) - %u",
+		    tx_bufs, tx_bufs ? *tx_bufs : NULL,
+		    rx_bufs, rx_bufs ? *rx_bufs : NULL, dfs);
+
+	ctx->current_tx = tx_bufs;
+	ctx->current_rx = rx_bufs;
+
+	if (*tx_bufs) {
+		ctx->tx_buf = (*tx_bufs)->buf;
+		ctx->tx_len = (*tx_bufs)->len/dfs;
+	} else {
+		ctx->tx_buf = NULL;
+		ctx->tx_len = 0;
+	}
+
+	if (*rx_bufs) {
+		ctx->rx_buf = (*rx_bufs)->buf;
+		ctx->rx_len = (*rx_bufs)->len/dfs;
+	} else {
+		ctx->rx_buf = NULL;
+		ctx->rx_len = 0;
+	}
+
+	SYS_LOG_DBG("current_tx %p, current_rx %p,"
+		    " tx buf/len %p/%u, rx buf/len %p/%u",
+		    ctx->current_tx, ctx->current_rx,
+		    ctx->tx_buf, ctx->tx_len, ctx->rx_buf, ctx->rx_len);
+}
+
+static ALWAYS_INLINE
+void spi_context_update_tx(struct spi_context *ctx, uint8_t dfs)
+{
+	if (!ctx->tx_len) {
+		return;
+	}
+
+	ctx->tx_len--;
+	if (!ctx->tx_len) {
```

Comment:
Can you please document these lines. They are not so obvious and it may help you to avoid questions from that are not as experienced as you are.